tests/validation_test.php

```diff
@@ -0,0 +1,161 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace local_headlessquiz;
+
+/**
+ * Headless quiz validation unit tests
+ *
+ * @package    local_headlessquiz
+ * @copyright  2022 Catalyst IT
+ * @author     Matthew Hilton <matthewhilton@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @covers \local_headlessquiz\api::get_headless_quiz
+ */
+class validation_test extends \advanced_testcase {
+
+    /**
+     * A selection of unsupported qtypes that allow unit testing (not all do).
+     */
+    private const UNSUPPORTED_TEST_QTYPES = [
+        'essay',
+        'description',
+        'numerical'
+    ];
+
+    public function setUp(): void {
+        $this->setAdminUser();
+        $this->resetAfterTest(true);
+
+        $this->dg = $this->getDataGenerator();
+        $this->qg = $this->dg->get_plugin_generator('mod_quiz');
+        $this->qsg = $this->dg->get_plugin_generator('core_question');
+
+        $this->user = $this->dg->create_user();
+        $this->course = $this->dg->create_course();
+        $this->dg->enrol_user($this->user->id, $this->course->id);
+        $this->quiz = $this->qg->create_instance(['course' => $this->course, 'grade' => 100.0, 'sumgrades' => 2,
+            'gradepass' => 50.0]);
+    }
+
+    public function test_user_doesnt_exist() {
+        // Call API with user ID that cannot exist.
+        $res = \local_headlessquiz\api::get_headless_quiz($this->quiz->cmid, -1);
+
+        // Should return validation error & no data.
+        $this->assertEquals(\local_headlessquiz\api::ERROR_VALIDATION, $res->error->type);
+        $this->assertEquals(get_string('error:validation:missinguser', 'local_headlessquiz'), $res->error->message);
+        $this->assertFalse(isset($res->data));
+    }
+
+    public function test_quiz_doesnt_exist() {
+        // Delete the quiz created in SetUp().
+        course_delete_module($this->quiz->cmid);
```

Comment:
I added a new test suite using `delete_course` and it works as expected, since the course module doesn't exist - are you referring to doing a broken  delete like `$DB->delete_record('course', ...` ??

Reply:
See https://github.com/catalyst/moodle-local_headlessquiz/blob/4dfed2e27c4b333054651767504f9aae4a37b806/tests/validation_test.php#L190-L203

Reply:
Yeah, you're right. As part of course deletion all its cms are detleted as well, so we get nice and clear error returned. `ERROR_UNKNOWN` will happen if the course module exists, but its course doesn't. As this shouldn't happen natively in Moodle I think it's fine to have `ERROR_UNKNOWN` returned. Resolving this thread :+1: